### PR TITLE
feat(project): add go.mod template and '--package-manager' flag to support go modules

### LINKED ISSUE #239: Add support for golang module function
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**What happened**:

Now, only configuration files for `dep`  dose nirvana command tool `nirvana` support to generate. But since [version 1.11](https://golang.org/doc/go1.11#modules) , Go has added preliminary support for a new concept called “modules” which will be the official solution of project dependency management.

**What you expected to happen**:

Provide an option for `dep` and  `mod` while user running `nirvana init ./myproject`

**How to reproduce it (as minimally and precisely as possible)**:

modify `nirvana/cmd/nirvana/project/init.go`:

1. add template method for mod file like `templateGopkg` method.
2. register the template. 

**Anything else we need to know?**:

### PR DESCRIPTION
**Pull request by MephistoMMM**
add manual for go module support (#239)

modify GoMod field of initOptions to Pkgmgr (#239)

modify "--mod" argument in manual project.md (#239)

fixbug: miss log when the value of pkgmrg is invalid (#239)

add target mod-reset-vendor (#239)

**What this PR does / why we need it**:

This PR add `pkgmgr`(*package manager*) argument to `nirvana init` command. Currently, The additional argument has two valid values - `dep` or `mod` , corresponding to pkg manager dep and go module function respectively.

I finally replaced `--mod` with `--pkgmgr` instead of `--vendor`. In my opinion, it's semantically more appropriate.

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #239

Reference to #

**Special notes for your reviewer**:

/assign @ddysher

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by kdada** [comment]
You need to fix build issues. https://travis-ci.org/caicloud/nirvana/builds/434464199#L563

You don't need to squash commits and add issue id in commit comments. Nirvana merge process will do it.

File: cmd/nirvana/project/init.go [hunk 1]
```diff
@@ -95,6 +101,7 @@ func (o *initOptions) Install(flags *pflag.FlagSet) {
 	flags.StringVar(&o.ImageSuffix, "image-suffix", "", "Docker image suffix")
 	flags.StringVar(&o.BuildImage, "build-image", "golang:latest", "Golang image for building the project")
 	flags.StringVar(&o.RuntimeImage, "runtime-image", "debian:jessie", "Docker base image for running the project")
+	flags.StringVar(&o.Pkgmgr, "pkgmgr", "dep", "Package manager of project, dep or mod")
```
Comment (kdada) [inline cmd/nirvana/project/init.go]: The `--pkgmgr` is not like other flags. `--package-manager` is better.

Comment (MephistoMMM) [reply cmd/nirvana/project/init.go]: ok

Comment (MephistoMMM) [reply cmd/nirvana/project/init.go]: I have modified it

File: cmd/nirvana/project/init.go [hunk 2]
```diff
@@ -709,7 +762,10 @@ build:
 	    $(CMD_DIR)/$${target};                                                         \
 	done
 
-container:
+mod-reset-vendor:
+	@$(shell [ -f go.mod ] && go mod vendor)
```
Comment (kdada) [inline cmd/nirvana/project/init.go]: Have you tested this line?

In my case,  this command always output:
```
$ make mod-rest-vendor
make: 'mod-rest-vendor' is up to date.
```

File: cmd/nirvana/project/init.go [hunk 3]
```diff
@@ -226,13 +234,31 @@ func (o *initOptions) templates(proj string) map[string]string {
...
 		"pkg/version/version.go":                 o.templateVersion(),
-		"Gopkg.toml":                             o.templateGopkg(),
+		pkgConfPath:                              templateGopkg(),
 		"Makefile":                               o.templateMakefile(),
 		project.DefaultProjectFileName:           o.templateProject(),
 		"README.md":                              o.templateReadme(),
 	}
 }
 
+func (o *initOptions) exTemplateGopkg() (string, func() string) {
```
Comment (kdada) [inline cmd/nirvana/project/init.go]: `exTemplateGopkg()` selects correct template for package manager. So `templatePackageManager()` may be a better method name.

File: cmd/nirvana/project/init.go [hunk 4]
```diff
@@ -638,6 +664,33 @@ required = ["github.com/caicloud/nirvana/utils/api"]
...
+//
+// module github.com/my/module/v3
+//
+// require (
+//     github.com/some/dependency v1.2.3
+//     github.com/another/dependency v0.1.0
+//     github.com/additional/dependency/v4 v4.0.0
+// )
+
+module {{ .ProjectName }}
```
Comment (kdada) [inline cmd/nirvana/project/init.go]: `ProjectName` is only the name of your project. But in the example, `module` is path to your project. Should this field be `ProjectPackage`?

Comment (MephistoMMM) [reply cmd/nirvana/project/init.go]: Yes, you're right.

File: cmd/nirvana/project/init.go (hunk 2 above)
Comment (kdada) [reply cmd/nirvana/project/init.go]: I've tested `@$(shell [ -f go.mod ] && echo 'something')`. But it alway outputs `make: 'mod-rest-vendor' is up to date.` even though I changed `go.mod`.

File: cmd/nirvana/project/init.go [hunk 5]
```diff
@@ -214,6 +221,7 @@ func (o *initOptions) directories(project string) []string {
 }
 
 func (o *initOptions) templates(proj string) map[string]string {
+	pkgConfPath, templateGopkg := o.templatePackageManager()
```
Comment (kdada) [inline cmd/nirvana/project/init.go]: Variable `templateGopkg` is not good. It may be a template of `mod`.

File: cmd/nirvana/project/init.go [hunk 6]
```diff
@@ -226,13 +234,31 @@ func (o *initOptions) templates(proj string) map[string]string {
...
 		"Makefile":                               o.templateMakefile(),
 		project.DefaultProjectFileName:           o.templateProject(),
 		"README.md":                              o.templateReadme(),
 	}
 }
 
+func (o *initOptions) templatePackageManager() (string, func() string) {
+	switch o.PackageManager {
+	case "mod":
+		return "go.mod", func() string {
```
Comment (kdada) [inline cmd/nirvana/project/init.go]: Why don't you just `return "go.mod", o. templateGomod `?

Comment (MephistoMMM) [reply cmd/nirvana/project/init.go]: 2333333333

File: cmd/nirvana/project/init.go (hunk 5 above)
Comment (MephistoMMM) [reply cmd/nirvana/project/init.go]: change it to` templatePkgmgr` ? Or, rename `templatePakcageManager` to other name, let `templateGopkg` be `templatePackageManager`.

File: cmd/nirvana/project/init.go (hunk 2 above)
Comment (MephistoMMM) [reply cmd/nirvana/project/init.go]: On my computer:

screen shot 2018-09-30 at 4 08 16 pm

screen shot 2018-09-30 at 4 09 00 pm

- - -

Would you like verbose logs as shown below ?

screen shot 2018-09-30 at 4 09 33 pm

File: cmd/nirvana/project/init.go (hunk 5 above)
Comment (kdada) [reply cmd/nirvana/project/init.go]: `templatePkgmgr` and `templatePackageManager` are ok.

Comment (MephistoMMM) [reply cmd/nirvana/project/init.go]: Ok.

File: cmd/nirvana/project/init.go (hunk 2 above)
Comment (kdada) [reply cmd/nirvana/project/init.go]: I've re-tested it. You are right.

**Comment by kdada** [comment]
Fix https://github.com/caicloud/nirvana/pull/240#pullrequestreview-160104350 and there are no other issues.

**Comment by MephistoMMM** [comment]
Tks , I have fixed it.

**Pull request merged**